### ISSUE TITLE
WP CLI theme install command hangs when domain is not pointed at host

### ISSUE THREAD

**Issue by gravyboat** [wp-cli/wp-cli#3401]
Information requested from Bug Report page:

```
OS Release: Ubuntu 15.10
PHP: 5.6.11
WP Version: 4.6.1
WP-CLI: 0.24.1
```

We've encountered an issue during theme installation with wp-cli running the following command where the command does not complete if the domain that the theme is being installed on does not point to the appropriate host:

```
/usr/local/bin/wp --path=/path_to_wordpress_install theme install --activate https://domain.s3.amazonaws.com/path/theme.zip
```

I've confirmed this connection issue via an strace on the associated php process that shows the system continually waiting on a file descriptor:

```
poll([{fd=5, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
clock_gettime(CLOCK_MONOTONIC, {3048477, 189836892}) = 0
clock_gettime(CLOCK_MONOTONIC, {3048477, 189886883}) = 0
clock_gettime(CLOCK_MONOTONIC, {3048477, 189936969}) = 0
```

Reviewing the associated file descriptor shows the following details:

```
php         341              user    5u     IPv4           34347517       0t0          TCP local_ip:PORT->184.168.221.55:http (ESTABLISHED)
```

This points at the DNS of the domain (a GoDaddy parked domain page) as the domain isn't currently pointed towards any sort of hosting. The process will continually wait on this with no way to proceed until it eventually seems to silently fail. Is there a way to work around this without modifying the domain's DNS, or am I missing something with how this works? I looked at the associated `theme` function and I don't see where it would be doing this sort of check, though my PHP is pretty bad. I've encountered this problem once before but didn't have enough details regarding the process to file an issue. Any insight you can provide regarding this issue would be appreciated. Is there a timeout being enforced on this command? Is there any way to bypass this 'check' and is wp-cli even doing such a check?

**Comment by danielbachhuber** [comment]
> We've encountered an issue during theme installation with wp-cli running the following command where the command does not complete if the domain that the theme is being installed on does not point to the appropriate host:
>
> [...]
> 
> This points at the DNS of the domain (a GoDaddy parked domain page) as the domain isn't currently pointed towards any sort of hosting. The process will continually wait on this with no way to proceed until it eventually seems to silently fail.

Some important qualifying questions:

* Does it happen for other commands too (e.g. installing a plugin), or just installing a theme?
* Does the issue reproduce consistently, or intermittently?
* Are you running custom code? Does the issue produce on a stock install of WordPress, or when you include `--skip-plugins` `--skip-themes`?
* Can you share the full output of the command?

> Is there a way to work around this without modifying the domain's DNS, or am I missing something with how this works? I looked at the associated `theme` function and I don't see where it would be doing this sort of check, though my PHP is pretty bad.

Yeah, I'm not quite sure why it would be doing a check of itself either. Seems suspicious.

**Comment by gravyboat** [comment]
* Does it happen for other commands too (e.g. installing a plugin), or just installing a theme? - This is unique to theme installations, other commands do not cause this http request to be created.
* Does the issue reproduce consistently, or intermittently? - Consistently. We've noticed it happen on several domains that just happened to be parked at GoDaddy parked pages, typically domains don't point elsewhere so while we can reproduce it, it is rare. If a domain is pointed to an actual hosted platform where the request will complete the process does not stall at this stage.
* Are you running custom code? Does the issue produce on a stock install of WordPress, or when you include --skip-plugins --skip-themes? - No custom code going on here.
* Can you share the full output of the command? - This is the full output of the command, it doesn't go past this point.

**Comment by danielbachhuber** [comment]
Ok.

Unfortunately, without clear directions on how to reproduce this issue locally, I can't help determine the source of the issue.

**Comment by gravyboat** [comment]
This issue should be reproducible by having a domain pointed to a provider (in this case GoDaddy, though it may occur elsewhere) where a domain is not actually hosted, so a generic parked page that doesn't have a wordpress installation. Then simply run `/usr/local/bin/wp --path=/path_to_wordpress_install theme install --activate https://domain.s3.amazonaws.com/path/theme.zip` and the issue should present itself where this command hangs and then silently fails.

**Comment by jmazzi** [comment]
@danielbachhuber is there any other information from us that could help narrow it down? This issue has been especially troubling for us.

**Comment by danielbachhuber** [comment]
> is there any other information from us that could help narrow it down?

Specific reproduction steps, of course.

> This issue should be reproducible by having a domain pointed to a provider (in this case GoDaddy, though it may occur elsewhere) where a domain is not actually hosted, so a generic parked page that doesn't have a wordpress installation. 

Using a WordPress site with an invalid domain `foobar.blog`, I can't reproduce the issue:

```
salty-wordpress ➜  themeinstalltest  wp core download
Downloading WordPress 4.6.1 (en_US)...
Using cached file '/home/vagrant/.wp-cli/cache/core/wordpress-4.6.1-en_US.tar.gz'...
Success: WordPress downloaded.
salty-wordpress ➜  themeinstalltest  wp core config --dbname=themeinstalltest --dbuser=root
Success: Generated 'wp-config.php' file.
salty-wordpress ➜  themeinstalltest  wp db create
Success: Database created.
salty-wordpress ➜  themeinstalltest  wp core install --url=foobar.blog --title=foobar
Success: WordPress installed successfully.
salty-wordpress ➜  themeinstalltest  wp theme install https://downloads.wordpress.org/theme/denim.0.3.zip --activate
Downloading install package from https://downloads.wordpress.org/theme/denim.0.3.zip...
Unpacking the package...
Installing the theme...
Theme installed successfully.
Activating 'denim'...
Success: Switched to 'Denim' theme.
```

If you can provide exact reproduction steps I can run locally that point to a problem within WP-CLI, then I can assess the issue and look into a fix. Otherwise, it's likely the problem exists independent of WP-CLI.

**Comment by itspriddle** [comment]
Hey @danielbachhuber, I did some digging and found some more information. This isn't actually a problem with `wp theme install` directly, but a combination of two things.

We believe that the GoDaddy domain that initially caused this issue is [tarpitting](https://en.wikipedia.org/wiki/Tarpit_(networking)) us. That's obviously not a WP-CLI issue, but it is the cause of the slowness and a side-effect of a bug (or feature?).

It seems that WP-CLI can randomly trigger wp-cron to fire (aside from `wp cron` commands due to [this](https://github.com/wp-cli/wp-cli/blob/f47271774d5337103a2c41c3b18b2071b5b8ba98/php/WP_CLI/Runner.php#L913). Is this by design? It seems unexpected to me, but I will admit may be missing something. As a work-around, we can explicitly `define( 'DOING_CRON', true )` in a script and `wp --require=script.php`, but we weren't sure if this is expected behavior.

So what's happening in our example, is that `wp theme install file.zip` (any `wp` command really) is sometimes running at just the right moment to trigger wp-cron and GoDaddy is tarpitting us resulting in a huge delay for the command to complete.

Let us know what you think, and thanks for the help! 

--

If you want to see this in action these are the steps I used (an example GoDaddy domain that can be used is montemorano.com, but it doesn't really matter):

1. Setup a bare WordPress 4.6.1 instance
2. In one terminal: `sudo ngrep -qt 'T' | grep wp-cron` (this was on OS X, maybe `ngrep -qt 'TCP'` for Linux)
3. In another terminal: `while true; do wp > /dev/null; done`
4. Back in the terminal running `ngrep`, you will eventually see output like `POST /wp-cron.php?doing_wp_cron=1473888795.2733130455017089843750 HTTP/1.1`

**Comment by danielbachhuber** [comment]
> It seems that WP-CLI can randomly trigger wp-cron to fire (aside from `wp cron` commands due to this. Is this by design? It seems unexpected to me, but I will admit may be missing something.

Yes, this is by design. Just like any web request can trigger WP Cron, any WP-CLI request can trigger WP Cron. I agree it can seem unexpected. Many ways of how WordPress works are unexpected :)

> As a work-around, we can explicitly `define( 'DOING_CRON', true )` in a script and `wp --require=script.php`, but we weren't sure if this is expected behavior.

I don't think this approach is very precise because sometimes plugins pay attention to the `DOING_CRON` constant. It would be better to define `ALTERNATE_WP_CRON=true` in some way that's only executed for WP-CLI requests. See #3104

> So what's happening in our example, is that `wp theme install file.zip` (any `wp` command really) is sometimes running at just the right moment to trigger wp-cron and GoDaddy is tarpitting us resulting in a huge delay for the command to complete.

Right, this makes sense to me.

**Comment by itspriddle** [comment]
Cool, thanks for the info!

**Comment by jmazzi** [comment]
@danielbachhuber would calling `remove_action( 'init', 'wp_cron' )` give us the same effect?
